### ISSUE TITLE
rpm_package silently fails to install packages with a `+` (plus) character in the version

### ISSUE THREAD

**Issue by irvingpop** [chef/chef#3671]
Very similar to the issue described in #3496 and fixed in #3503 

This line of code: https://github.com/chef/chef/blob/master/lib/chef/provider/package/rpm.rb#L64
is proving brittle because I've found yet another character that it misses:
http://rubular.com/r/kdkr3EYxOY

Using Chef 12.4.1:

```
chef:recipe (12.4.1)> run_chef
[2015-07-15T17:38:39+00:00] INFO: Processing rpm_package[/tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm] action install ((irb#1) line 1)
[2015-07-15T17:38:39+00:00] DEBUG: Providers for generic rpm_package resource enabled on node include: [Chef::Provider::Package::Rpm]
[2015-07-15T17:38:39+00:00] DEBUG: Provider for action install on resource rpm_package[/tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm] is Chef::Provider::Package::Rpm
[2015-07-15T17:38:39+00:00] DEBUG: No package source specified, but /tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm exists on the filesystem, copying to package source
[2015-07-15T17:38:39+00:00] DEBUG: rpm_package[/tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm] checking rpm status
chef-server-core 12.2.0+20150713220422-1.el6
[2015-07-15T17:38:39+00:00] DEBUG: rpm_package[/tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm] checking install state
package /tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm is not installed
[2015-07-15T17:38:39+00:00] DEBUG: rpm_package[/tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm] /tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm not installed, installing 
[2015-07-15T17:38:39+00:00] DEBUG: rpm_package[/tmp/ecm_cache/chef-server-core-12.2.0+20150713220422-1.el6.x86_64.rpm] is already installed - nothing to do
 => true 
```

Before I go make a patch that adds `+` - is there a definitive guide for which characters are acceptable in an RPM version?

cc: @chef/client-core @chef/engineering-services

**Comment by lamont-granquist** [comment]
I'm fairly certain such a document does exist.

**Comment by yzl** [comment]
@irvingpop While there is undoubtedly an official doc somewhere, I couldn't find one.  I tend to look to the rpm doc, the Fedora standard, and the code:

http://fedoraproject.org/wiki/Packaging%3aNamingGuidelines (although we do use ~'s)
http://blog.jasonantman.com/2014/07/how-yum-and-rpm-compare-versions/ (useful for listing out places in code where the version data matters)
http://www.rpm.org/max-rpm/s1-rpm-inside-tags.html
http://www.rpm.org/max-rpm/ch-rpm-file-format.html

**Comment by lamont-granquist** [comment]
Can we invert this problem and instead of doing any validation at all just figure out what delimits the version number and accept any characters that are there?

**Comment by coderanger** [comment]
:+1: to @lamont-granquist, just faced the same thing with Python package version numbers, gave up and used `/[^\s,]+/`.

**Comment by lamont-granquist** [comment]
So according to wikipedia we have this:

```ruby
<name>-<version>-<release>.<architecture>.rpm
```

So my first stab looks like this:

```ruby
(.*)\-(.*?\-.*?)\..*?\.rpm
```

You can throw an '\A\z' around it and/or replace the '.'s with '\s'.  The first matcher should be greedy and slurp up every character it can and needs to have a '-' to the right of it, that is the name.  The second matcher is the version and must have exactly one '-' in it (it must have at least one '-' in it, but the previous greedy match will slurp up all but the last one), and be delimited on the right by a '.'

Then for .arch.rpm I made the arch non-greedy so that .foo-bar.baz.rpm wouldn't be accidentally slurped into the arch, although i don't know that will ever matter.

**Comment by lamont-granquist** [comment]
pingles?  anyone want to PR this?

**Comment by irvingpop** [comment]
Hey @lamont-granquist I have some cycles to work on this.    Thinking this through a bit:

* we shell out like so: `rpm -qp --queryformat '%{NAME} %{VERSION}-%{RELEASE}\n'` - is there any reason we couldn't trust these results and report back the two parts?
* if so, the regex could be as simple as `^(\S+)\s(\S+)$` [rubular link](http://rubular.com/r/j4U5gLGkpx)
* alternative is to be less trusting of the name, but greedy with the version like `^([\w\d+_.-]+)\s(\S+)$` [rubular link](http://rubular.com/r/Ay446zs97V)
* reading the links that @yzl posted - it seems like any code that tries to understand the version as Semver will fail on one edge case or another

**Comment by lamont-granquist** [comment]
should be fixed in both rpm_package by #3985 and yum_package localinstalls by #5614
